### PR TITLE
feat: display Yearn and uniV2 transactions via generic contract handling component

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

- Adds a generic `TransactionContract` component that handles contract interaction transactions, including Yearn vault and UniV2 liquidity interactions.
- Makes required package bumps

## Future improvements

- It'll require additional data from the unchained parser, but next, I'd like to add the contract that the transaction interacted within the expanded view.
- Add ZRX ABI and better support associated transactions
- Add a generic, fallback, transaction renderer

## Notes for the reviewer

- I'm checking with the product team to confirm the final UI
- Has a dependency on https://github.com/shapeshift/lib/pull/405

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/1068

## Risk

Minimal - it doesn't touch existing code, and at worst, the new functionality won't work as promised.

## Testing

- Check that UniSwap add and remove liquidity transactions display
- Check that Yearn approve, deposit, and withdraw transactions display

## Screenshots (if applicable)

Screen Shot 2022-02-28 at 4 41 20 pm
Screen Shot 2022-02-28 at 4 41 30 pm
Screen Shot 2022-02-28 at 4 42 24 pm

### CODE REVIEW

**Comment by 0xdef1cafe** [comment]
@0xApotheosis unblocked :)

**Comment by 0xApotheosis** [comment]
Swapped `FaSignature` with `FaThumbsUp` as per suggestion from @reallybeard in https://github.com/shapeshift/web/pull/1076/commits/15ca73fc20e61ad0386f0292f20bc13a0ab04fc2

Screen Shot 2022-03-01 at 10 50 17 am

File: src/hooks/useTxDetails/useTxDetails.ts [hunk 1]
```diff
@@ -25,21 +35,37 @@ export interface TxDetails {
...
-export const getSellTx = (tx: Tx) =>
-  !!tx.tradeDetails ? tx.transfers.find(t => t.type === chainAdapters.TxType.Send) : undefined
+export const getBuyTx = (tx: Tx) => tx.transfers.find(t => t.type === chainAdapters.TxType.Receive)
+export const getSellTx = (tx: Tx) => tx.transfers.find(t => t.type === chainAdapters.TxType.Send)
 
 export const useTxDetails = (txId: string, activeAsset?: Asset): TxDetails => {
   const tx = useSelector((state: ReduxState) => selectTxById(state, txId))
+  const method = tx.data?.method
+
+  const isSupportedContract = method ? SUPPORTED_CONTRACT_METHODS.has(method) : false
```
Comment (0xdef1cafe) [inline src/hooks/useTxDetails/useTxDetails.ts:49]: this little dude could move out above

**Review by 0xdef1cafe — APPROVED** [review]
approved w/ comments

File: src/hooks/useTxDetails/useTxDetails.ts (hunk 1 above)
Comment (0xApotheosis) [reply src/hooks/useTxDetails/useTxDetails.ts]: Moved out in https://github.com/shapeshift/web/pull/1076/commits/066dedd1f599ad8d3f4f26daf2a2b06de1e097a6.

File: src/components/Transactions/TransactionContract.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+          <Row variant='vertical' hidden={!txDetails.tradeTx}>
+            <Row.Label>
+              <Text translation={'transactionRow.amount'} />
+            </Row.Label>
+            <Row.Value>
+              <Amount.Crypto
+                value={fromBaseUnit(
+                  txDetails.sellTx?.value ?? '0',
+                  txDetails.sellAsset?.precision ?? 18
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Transactions/TransactionContract.tsx:117]: We use this pattern widely across components, I'll do a follow-up PR that fixes all instances.

Made an issue to track the follow-up: https://github.com/shapeshift/web/issues/1085

File: src/assets/translations/messages-en.json [hunk 3]
```diff
@@ -190,7 +190,21 @@
...
     "amount": "Amount",
-    "for": "for"
+    "for": "for",
+    "parser": {
+      "yearn": {
+        "deposit": "Deposit",
+        "approve": "Approve",
+        "withdraw": "Withdraw"
+      },
+      "uniV2": {
```
Comment (gomesalexandre) [inline src/assets/translations/messages-en.json:200]: Non-blocking: What about using a `methods` object to encapsulate these?

File: src/components/Transactions/TransactionContract.tsx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                maximumFractionDigits={6}
+              />
+            </Row.Value>
+          </Row>
+          <Row variant='vertical'>
+            <Row.Label>
+              <Text translation='transactionRow.status' />
+            </Row.Label>
+            <Row.Value textAlign='left'>
+              {txDetails.tx.status === chainAdapters.TxStatus.Confirmed && (
```
Comment (gomesalexandre) [inline src/components/Transactions/TransactionContract.tsx:154]: Non-blocking, but three conditional renders like this depending on a pattern match is a call for a render function that could use a switch case on `txDetails.tx.status`

File: src/components/Transactions/TransactionContract.tsx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                overflow='hidden'
+                flex={1}
+                textOverflow='ellipsis'
+                maxWidth='60%'
+                lineHeight='1'
+                whiteSpace='nowrap'
+                mb={2}
+                translation={[
+                  `transactionRow.parser.${txDetails.tx.data?.parser}.${txDetails.tx.data?.method}`,
+                  { symbol: '' }
```
Comment (gomesalexandre) [inline src/components/Transactions/TransactionContract.tsx:63]: There is no such interpolation in parser translations. I guess it can be removed?

**Review by gomesalexandre — APPROVED** [review]
From a first look at implementation and testing it out with the whole DeFi depositting/withdrawing flow, this looks good. Approve, deposit and withdraw rows are all working as expected (tested on USDC pool).

**Comment by 0xApotheosis** [comment]
Ty @gomesalexandre, I'll do a fast-follow PR re: your comments 🤙

File: src/components/Transactions/TransactionContract.tsx (hunk 5 above)
Comment (0xApotheosis) [reply src/components/Transactions/TransactionContract.tsx:63]: There is indeed such an interpolation, e.g. `transactionRow.parser.yearn.deposit` - unless I've misunderstood your comment?

File: src/hooks/useTxDetails/useTxDetails.ts [hunk 6]
```diff
@@ -25,21 +35,38 @@ export interface TxDetails {
   precision: number
   explorerTxLink: string
   explorerAddressLink: string
+  direction: 'in-place' | 'outbound' | 'inbound'
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/hooks/useTxDetails/useTxDetails.ts:38]: Addressed in https://github.com/shapeshift/web/pull/1096

**Pull request merged**